### PR TITLE
WELD-2268 implement trimmed bean archive.

### PR DESCRIPTION
**Pull request by tremes**

### CODE REVIEW

File: impl/src/main/java/org/jboss/weld/bootstrap/BeanDeployment.java [hunk 1]
```diff
@@ -279,4 +286,12 @@ public String toString() {
         return builder.toString();
     }
 
+    public void trim() {
+        for (SlimAnnotatedTypeContext<?> ctx : getBeanDeployer().getEnvironment().getAnnotatedTypes()) {
+            if (!AnnotatedTypes.hasBeanDefiningAnnotation(ctx.getAnnotatedType())) {
+                getBeanDeployer().getEnvironment().removeAnnotatedType(ctx);
```
Comment (tremes) [inline impl/src/main/java/org/jboss/weld/bootstrap/BeanDeployment.java]: It's unmodifiable set.

Comment (mkouba) [inline impl/src/main/java/org/jboss/weld/bootstrap/BeanDeployment.java]: Then we should probably perform the "trim" operation "inside" the `BeanDeployerEnvironment`?

Comment (tremes) [inline impl/src/main/java/org/jboss/weld/bootstrap/BeanDeployment.java]: Yes would make sense.

File: impl/src/main/java/org/jboss/weld/bootstrap/BeanDeployment.java [hunk 2]
```diff
@@ -250,9 +251,13 @@ public void createBeans(Environment environment) {
         }
         beanDeployer.addBuiltInBean(new RequestContextControllerBean(beanManager));
 
+        if (beanDeploymentArchive.getBeansXml() != null) {
```
Comment (mkouba) [inline impl/src/main/java/org/jboss/weld/bootstrap/BeanDeployment.java]: `if (beanDeploymentArchive.getBeansXml() != null && beanDeploymentArchive.getBeansXml().isTrimmed())` ?

File: impl/src/main/java/org/jboss/weld/bootstrap/BeanDeployerEnvironment.java [hunk 3]
```diff
@@ -406,4 +408,12 @@ public void cleanup() {
     public Set<Type> getNewBeanTypes() {
         return Collections.unmodifiableSet(newBeanTypes);
     }
+
+    public void trim() {
+        for (Iterator<SlimAnnotatedTypeContext<?>> iterator = annotatedTypes.iterator(); iterator.hasNext(); ) {
+            if (!AnnotatedTypes.hasBeanDefiningAnnotation(iterator.next().getAnnotatedType())) {
```
Comment (mkouba) [inline impl/src/main/java/org/jboss/weld/bootstrap/BeanDeployerEnvironment.java]: And what about pseudo scopes? E.g. `javax.inject.Singleton`.

File: impl/src/main/java/org/jboss/weld/util/AnnotatedTypes.java [hunk 4]
```diff
@@ -56,6 +65,13 @@
  */
 public class AnnotatedTypes {
 
+    private static List<Class<? extends Annotation>> beanDefiningAnnotations = ImmutableList.of(
```
Comment (mkouba) [inline impl/src/main/java/org/jboss/weld/util/AnnotatedTypes.java]: The name should be BEAN_DEFINING_ANNOTATIONS.

File: impl/src/main/java/org/jboss/weld/util/AnnotatedTypes.java [hunk 5]
```diff
@@ -56,6 +65,13 @@
...
+            Dependent.class, RequestScoped.class, ConversationScoped.class, SessionScoped.class, ApplicationScoped.class,
+            javax.interceptor.Interceptor.class, javax.decorator.Decorator.class,
+            Model.class);
+
+    private static List<Class<? extends Annotation>> metaAnnotations = ImmutableList.of(Stereotype.class, NormalScope.class);
```
Comment (mkouba) [inline impl/src/main/java/org/jboss/weld/util/AnnotatedTypes.java]: The name should be META_ANNOTATIONS.

**Comment by mkouba** [comment]
Pls rebase ;-)

File: impl/src/main/java/org/jboss/weld/xml/XmlSchema.java [hunk 6]
```diff
@@ -20,7 +20,7 @@
 
 public enum XmlSchema {
 
-    CDI10("beans_1_0.xsd", BeanManager.class.getClassLoader()), CDI11("beans_1_1.xsd", BeanManager.class.getClassLoader()), WELD11("weld_1_1.xsd", BeansXmlParser.class.getClassLoader());
+    CDI20("beans_2_0.xsd", BeanManager.class.getClassLoader()), WELD11("weld_1_1.xsd", BeansXmlParser.class.getClassLoader());
```
Comment (mkouba) [inline impl/src/main/java/org/jboss/weld/xml/XmlSchema.java]: We can't remove `beans_1_0.xsd` -> it has a different namespace and we would like to support old versions of `beans.xml`.

**Comment by mkouba** [comment]
Merged, thanks!

**Pull request closed without merging**